base36.go

```diff
@@ -0,0 +1,154 @@
+/*
+
+Package base36 provides a reasonably fast implementation of a binary base36 codec.
+
+*/
+package base36
+
+// Simplified code based on https://godoc.org/github.com/mr-tron/base58
+// which in turn is based on https://github.com/trezor/trezor-crypto/commit/89a7d7797b806fac
+
+import (
+	"fmt"
+)
+
+const UcAlphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+const LcAlphabet = "0123456789abcdefghijklmnopqrstuvwxyz"
+const maxDigitOrdinal = byte('z')
+const maxDigitValueB36 = 35
+
+var revAlphabet [maxDigitOrdinal + 1]byte
+
+func init() {
+	for i := range revAlphabet {
+		revAlphabet[i] = maxDigitValueB36 + 1
+	}
+	for i, c := range UcAlphabet {
+		revAlphabet[byte(c)] = byte(i)
+		if c > '9' {
+			revAlphabet[byte(c)+32] = byte(i)
+		}
+	}
+}
+
+// EncodeToStringUc encodes the given byte-buffer as base36 using [0-9A-Z] as
+// the digit-alphabet
+func EncodeToStringUc(b []byte) string { return encode(b, UcAlphabet) }
+
+// EncodeToStringLc encodes the given byte-buffer as base36 using [0-9a-z] as
+// the digit-alphabet
+func EncodeToStringLc(b []byte) string { return encode(b, LcAlphabet) }
+
+func encode(inBuf []byte, al string) string {
+
+	// As a polar opposite to the base58 implementation, using a uint32 here is
+	// significantly slower
+	var carry uint
+
+	var encIdx, valIdx, zcnt, high int
+
+	inSize := len(inBuf)
+	for zcnt < inSize && inBuf[zcnt] == 0 {
+		zcnt++
+	}
+
+	// Really this is log(256)/log(36) or 1.55, but integer math is easier
+	// Use 2 as a constant and just overallocate
+	encSize := (inSize - zcnt) * 2
+
+	// Allocate one big buffer up front
+	// Note: pools *DO NOT* help, the overhead of zeroing the val-half (see below)
+	// kills any performance gain to be had
+	outBuf := make([]byte, (zcnt + encSize*2))
+
+	// use the second half for the temporary numeric buffer
+	val := outBuf[encSize+zcnt:]
+
+	high = encSize - 1
+	for _, b := range inBuf[zcnt:] {
+		valIdx = encSize - 1
+		for carry = uint(b); valIdx > high || carry != 0; valIdx-- {
+			carry += uint((val[valIdx])) * 256
```

Comment:
nit: I'd use uint64 or uint32 to so we don't run into weird overflow issues.

Reply:
Will fix. I have a very corner case bug it seems, still working on orting the tests. Will merge once this is all resolved.